ding/policy/ibc.py

```diff
@@ -0,0 +1,165 @@
+        return 'ebm', ['ding.model.template.ebm']
+    
+    def _init_learn(self):
+        optim_cfg = self._cfg.learn.optim
+        self._optimizer = torch.optim.Adam(
+            self._model.parameters(),
+            lr=optim_cfg.learning_rate,
+            weight_decay=optim_cfg.weight_decay,
+            betas=(optim_cfg.beta1, optim_cfg.beta2),
+        )
+        self._stochastic_optimizer: StochasticOptimizer = \
+            create_stochastic_optimizer(self._cfg.model.stochastic_optim)
+        self._learn_model = model_wrap(self._model, 'base')
+        self._learn_model.reset()
+    
+    def _forward_learn(self, data):
+        if not isinstance(data, dict):
+            data = default_collate(data)
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._learn_model.train()
+
+        loss_dict = dict()
+
+        # obs: (B, O)
+        # action: (B, A)
+        obs, action = data['obs'], data['action']
+
+        # (B, N, O), (B, N, A)
+        obs, negatives = self._stochastic_optimizer.sample(obs, self._learn_model)
+
+        # (B, N+1, A)
+        targets = torch.cat([action.unsqueeze(dim=1), negatives], dim=1)
+        # (B, N+1, O)
+        obs = torch.cat([obs[:, :1], obs], dim=1)
+
+        permutation = torch.rand(targets.shape[0], targets.shape[1]).argsort(dim=1)
+        targets = targets[torch.arange(targets.shape[0]).unsqueeze(-1), permutation]
+
+        ground_truth = (permutation == 0).nonzero()[:, 1].to(self._device)
+
+        # (B, N+1) for ebm
+        # (B, N+1, A) for autoregressive ebm
+        energy = self._learn_model.forward(obs, targets)
+
+        logits = -1.0 * energy
+        if isinstance(self._stochastic_optimizer, AutoRegressiveDFO):
+            # autoregressive case
+            # (B, A)
+            ground_truth = unsqueeze_repeat(ground_truth, logits.shape[-1], -1)
+        loss = F.cross_entropy(logits, ground_truth)
+        loss_dict['ebm_loss'] = loss.item()
+
+        if isinstance(self._stochastic_optimizer, MCMC):
+            grad_penalty = self._stochastic_optimizer.grad_penalty(obs, targets, self._learn_model) 
+            loss += grad_penalty
+            loss_dict['grad_penalty'] = grad_penalty.item()
+        loss_dict['total_loss'] = loss.item()
+
+        self._optimizer.zero_grad()
+        loss.backward()
+        self._optimizer.step()
+
+        return loss_dict
+
+    def _monitor_vars_learn(self):
+        if isinstance(self._stochastic_optimizer, MCMC):
+            return ['total_loss', 'ebm_loss', 'grad_penalty']
+        else:
+            return ['total_loss', 'ebm_loss']
+
+    def _init_eval(self):
+        self._eval_model = model_wrap(self._model, wrapper_name='base')
+        self._eval_model.reset()
+
+    def _forward_eval(self, data: dict) -> dict:
+        tensor_input = isinstance(data, torch.Tensor)
+        if tensor_input:
+            data = default_collate(list(data))
```

Comment:
why not just pass

Reply:
Changed. I reused and copied some of the infrastructure codes from `bc.py` and did not realize they do not really make sure.